prov/efa/src/efa_mr.c

```diff
@@ -547,6 +550,163 @@ static struct ibv_mr *efa_mr_reg_ibv_mr(struct efa_mr *efa_mr, struct fi_mr_attr
+	struct fid_mr *existing_mr_fid;
+	struct efa_mr *existing_mr;
+	CUmemorytype cuda_memory_type;
+	int err;
+
+	mr_attr->requested_key = efa_mr->mr_fid.key;
+	ofi_genlock_lock(&efa_mr->domain->util_domain.lock);
+	err = ofi_mr_map_insert(&efa_mr->domain->util_domain.mr_map, mr_attr,
+				&efa_mr->mr_fid.key, &efa_mr->mr_fid);
+	ofi_genlock_unlock(&efa_mr->domain->util_domain.lock);
+	if (!err)
+		return 0;
+
+	/* There is a special error that we can recover from, which is:
+	 * 1. error code is FI_ENOKEY, which means there is already a MR in mr_map with the same key
+	 * 2. that MR is for a CUDA memory region
+	 * 3. that CUDA memory region has been freed.
+	 *
+	 * This situation can happen because the combination of the following 3 things:
+	 * 1. cuda memory uses memory cache.
+	 * 2. cuda memory cache's monitor is not monitoring call to cudaFree().
+	 * 3. kernel released the kernel space of memory registration when cudaFree() is called on a registered cuda memory.
+	 *
+	 * Therefore, when application call cudaFree() on a registered CUDA memory region, the EFA kernel module
+	 * device released the kernel space memory region, made the key available for reuse.
+	 * However, libfabric is not aware of the incident, and kept the user space memory registration in its
+	 * MR cache and MR map.
+	 *
+	 * After we implement cuda memhook monitor (which monitors call to cudaFree() and remove dead region from MR cache),
+	 * we should NOT encounter this special situation any more. At that time, this function should be removed.
+	 */
+	if (err != -FI_ENOKEY) {
+		/* no way we can recover from this error, return error code */
+		EFA_WARN(FI_LOG_MR,
+			"Unable to add MR to map. errno: %d errmsg: (%s) key: %ld buff: %p hmem_iface: %d len: %zu\n",
+			err,
+			fi_strerror(-err),
+			efa_mr->mr_fid.key,
+			mr_attr->mr_iov->iov_base,
+			mr_attr->iface,
+			mr_attr->mr_iov->iov_len);
+		return err;
+	}
+
+	existing_mr_fid = ofi_mr_map_get(&efa_mr->domain->util_domain.mr_map, efa_mr->mr_fid.key);
+	assert(existing_mr_fid);
+	existing_mr = container_of(existing_mr_fid, struct efa_mr, mr_fid);
+
+	if (existing_mr->peer.iface != FI_HMEM_CUDA) {
+		/* no way we can recover from this situation, return error code */
+		EFA_WARN(FI_LOG_DOMAIN, "key %ld already assigned to buffer: %p hmem_iface: %d length: %ld\n",
+			 existing_mr->mr_fid.key,
+			 existing_mr->ibv_mr->addr,
+			 existing_mr->peer.iface,
+			 existing_mr->ibv_mr->length);
+		return -FI_ENOKEY;
+	}
+
+	/* According to CUDA document, the return code of ofi_cuPointerGetAttribute() being CUDA_ERROR_INVALID_VALUE
+	 * means the exising_mr's pointer is NOT allocated by, mapped by or registered with CUDA. In this context,
+	 * it means existing_mr's address has been de-allocated (which is the only situation we can recover from).
+	 */
+	err = ofi_cuPointerGetAttribute(&cuda_memory_type, CU_POINTER_ATTRIBUTE_MEMORY_TYPE,
+					(CUdeviceptr)existing_mr->ibv_mr->addr);
+	if (err != CUDA_ERROR_INVALID_VALUE) {
```

Comment:
We discussed and there needs to be an additional check here that compares buffer id as there is a valid situation where the device re-uses the virutal address for a new allocation.

Reply:
In latest revision, I added the check for buffer ID. please take a look